### PR TITLE
fix: Append CodeSha256 Value to Description

### PR DESCRIPTION
**Pull request by GavinZZ**
### Description of changes
Append the value of CodeSha256 to end of Lambda Function description. This would cause an update to the existing users who use codesha256 but since it will always create a new lambda version, it wouldn't cause any downtime for users. 

```
When you deploy a new version of an AWS Lambda function, it does not cause downtime for the existing function. AWS Lambda is designed to allow seamless deployments without affecting the current running version
```

Additionally add `DeletionPolicy` and `UpdateReplacePolicy` to an integ test.

### Description of how you validated changes
All tests passed.

### Checklist

- [ ] Adheres to the [development guidelines](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#development-guidelines)
- [ ] Add/update [transform tests](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#unit-testing-with-multiple-python-versions)
    - [ ] Using correct values
    - [ ] Using wrong values
- [ ] Add/update [integration tests](https://github.com/aws/serverless-application-model/blob/develop/INTEGRATION_TESTS.md)

### Examples?

Please reach out in the comments if you want to add an example. Examples will be 
added to `sam init` through [aws/aws-sam-cli-app-templates](https://github.com/aws/aws-sam-cli-app-templates).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samtranslator/model/sam_resources.py [hunk 1]
```diff
@@ -290,6 +290,13 @@ def to_cloudformation(self, **kwargs):  # type: ignore[no-untyped-def]
                         self.logical_id,
                         "AutoPublishCodeSha256 must be a string",
                     )
+                # Lambda doesn't create a new version if the code in the unpublished version is the same as the
+                # previous published version. To handle scenario users only update content in 'CodeUri', CloudFormation
```
Comment (paulhcsun) [inline samtranslator/model/sam_resources.py:294]: nit: "To handle this scenario, users..."

File: samtranslator/model/sam_resources.py [hunk 2]
```diff
@@ -290,6 +290,13 @@ def to_cloudformation(self, **kwargs):  # type: ignore[no-untyped-def]
...
+                # won't be able to detect any change in lambda function in the template, which would cause deployment
+                # failure. Append to description to handle this issue
+                description = intrinsics_resolver.resolve_parameter_refs(self.Description)
```
Comment (ssenchenko) [inline samtranslator/model/sam_resources.py:297]: What if `Description` uses other intrinsics which we fail to resolve?
Would it be safer to use `!Join` intrinsic to add `code_sha_256`?

File: samtranslator/model/sam_resources.py (hunk 1 above)
Comment (GavinZZ) [reply samtranslator/model/sam_resources.py:294]: I don't think that's what I'm trying to convey. I want to say to handle the scenario that users only updated content in `CodeUri`, ..... Let me rephrase it.

**Review by ssenchenko — COMMENTED** [review]
Do I understand it correctly that you're saying that updating `Description`will cause redeployment?

File: samtranslator/model/sam_resources.py (hunk 2 above)
Comment (GavinZZ) [reply samtranslator/model/sam_resources.py:297]: Sorry my description is not accurate. I meant it will cause an update to users who use CodeSha256 in their template even when there's no change made.

Comment (GavinZZ) [reply samtranslator/model/sam_resources.py:297]: If `Description` uses other intrinsics, we would not resolve it here. `!Join` is definitely a viable solution. The reason I didn't use `!Join` is that I feel like it's an overkill since not many users would use `Description` with other intrinsics and use `AutoPublishCodeSha256` at the same time. What do you think?

Comment (ssenchenko) [reply samtranslator/model/sam_resources.py:297]: I agree that it sounds like an overkill and in 99.90% there will be no need in `Join` but I was surprised when I got a ticket from users using intrinsics on Name in GraphQL to override LogicalId. It'd say it's better to be safe.

**Comment by SimonCMoore** [comment]
How about adding a test for when there is already a description and one for a description with an intrinsic in it?

**Comment by GavinZZ** [comment]
Added

**Review by ssenchenko — APPROVED** [review]

**Review by paulhcsun — APPROVED** [review]

**Comment by ramosbugs** [comment]
Btw, this was a breaking change since this caused some of our Lambda function descriptions to now be longer than the max length of 256, and deployment failed. Now we have to shorten our descriptions to compensate... something to be aware of for the future.
